File: algorithms/2-add-two-numbers/add_two_numbers.go
package add_two_numbers

type ListNode struct {
	Val  int
	Next *ListNode
}

// https://leetcode.com/problems/add-two-numbers/
func addTwoNumbersIter(l1 *ListNode, l2 *ListNode) *ListNode {
	current := &ListNode{}
	head := current
	carry := 0
	for {
		sum := carry

		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		if sum >= 10 {
			carry = sum / 10
			sum %= 10
		} else {
			carry = 0
		}

		current.Val = sum
		if l1 != nil || l2 != nil || carry != 0 {
			current.Next = &ListNode{}
			current = current.Next
		} else {
			break
		}
	}

	return head
}

func addTwoNumbersRecur(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil && l2 == nil {
		// recursion has bottomed out, return back up the chain
		return nil
	}

	sum := 0
	if l1 != nil {
		sum += l1.Val
	}
	if l2 != nil {
		sum += l2.Val
	}

	if sum >= 10 {
		carry := sum / 10
		sum %= 10
		if l1.Next != nil {
			l1.Next.Val += carry
		} else {
			l1.Next = &ListNode{Val: carry, Next: nil}
		}
	}

	// prevent calling .Next on nil node
	var l1Next, l2Next *ListNode
	if l1 != nil {
		l1Next = l1.Next
	}
	if l2 != nil {
		l2Next = l2.Next
	}

	next := addTwoNumbersRecur(l1Next, l2Next)

	return &ListNode{Val: sum, Next: next}
}

File: algorithms/2-add-two-numbers/add_two_numbers.py
from typing import Optional


class ListNode:
    """Definition for singly-linked list"""

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    """https://leetcode.com/problems/add-two-numbers/"""

    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if l1 and not l2:
            return l1
        elif not l1 and l2:
            return l2
        elif not l1 and not l2:
            return None

        sum: int = l1.val + l2.val
        # Let the rest of the recursion run, which returns the rest of the number chain,
        # not including any carrying we may need to do for two digits with sum >= 10
        next: Optional[ListNode] = self.addTwoNumbers(l1.next, l2.next)
        if sum >= 10:
            # sum can't be more than 18 (from adding two nines),
            # so we only need to modulo once to get back to a single digit
            sum %= 10
            # The next digit (which is one tens place higher) needs to be increased by 1
            # from whatever came back from the rest of the recursion.
            # If the next digit came back >= 10 (only possible case is next.Val = 9),
            # then the recursive call will hit this code path again and correct it.
            next: Optional[ListNode] = self.addTwoNumbers(
                next,
                ListNode(val=1, next=None),
            )
        return ListNode(val=sum, next=next)
